presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -561,27 +573,64 @@ public static long stNumGeometries(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
-        // https://github.com/Esri/geometry-api-java/issues/176
-        // https://github.com/Esri/geometry-api-java/issues/177
-        OGCGeometry leftGeometry = deserialize(left);
-        validateType("ST_Union", leftGeometry, EnumSet.of(POINT, MULTI_POINT, LINE_STRING, MULTI_LINE_STRING, POLYGON, MULTI_POLYGON));
-        if (leftGeometry.isEmpty()) {
-            return right;
+        return stUnion(ImmutableList.of(left, right));
+    }
+
+    @Description("Returns a geometry that represents the point set union of the input geometries.")
+    @ScalarFunction("geometry_union")
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice geometryUnion(@SqlType("array(" + GEOMETRY_TYPE_NAME + ")") Block input)
+    {
+        return stUnion(getGeometrySlicesFromBlock(input));
+    }
+
+    private static Slice stUnion(Iterable<Slice> slices)
+    {
+        // The current state of Esri/geometry-api-java does not allow support for multiple dimensions being
+        // fed to the union operator without dropping the lower dimensions:
+        // https://github.com/Esri/geometry-api-java/issues/199
+        // When operating over a collection of geometries, it is more efficient to reuse the same operator
+        // for the entire operation.  Therefore, split the inputs and operators by dimension, and then union
+        // each dimension's result at the end.
+        ListeningGeometryCursor[] cursorsByDimension = new ListeningGeometryCursor[NUMBER_OF_DIMENSIONS];
+        GeometryCursor[] operatorsByDimension = new GeometryCursor[NUMBER_OF_DIMENSIONS];
+
+        setAll(cursorsByDimension, i -> new ListeningGeometryCursor());
+        setAll(operatorsByDimension, i -> OperatorUnion.local().execute(cursorsByDimension[i], null, null));
+
+        for (Slice slice : slices) {
+            // Ignore null inputs
+            if (slice.getInput().available() == 0) {
+                continue;
+            }
+
+            for (OGCGeometry geometry : flattenCollection(deserialize(slice))) {
+                int dimension = geometry.dimension();
+                cursorsByDimension[dimension].tick(geometry.getEsriGeometry());
+                operatorsByDimension[dimension].tock();
+            }
         }
 
-        OGCGeometry rightGeometry = deserialize(right);
-        validateType("ST_Union", rightGeometry, EnumSet.of(POINT, MULTI_POINT, LINE_STRING, MULTI_LINE_STRING, POLYGON, MULTI_POLYGON));
-        if (rightGeometry.isEmpty()) {
-            return left;
+        List<OGCGeometry> outputs = new ArrayList<>();
+        for (GeometryCursor operator : operatorsByDimension) {
+            OGCGeometry unionedGeometry = createFromEsriGeometry(operator.next(), null);
+            if (unionedGeometry != null) {
+                outputs.add(unionedGeometry);
+            }
         }
 
-        return serialize(leftGeometry.union(rightGeometry));
+        if (outputs.isEmpty()) {
```

Comment:
Does this happen when all geometries are empty? If so, this appears to be a change in semantics. The old version of the code seemed to do:

```
if (rightGeometry.isEmpty()) {
    return left;
}
```